File: backend/packages/ai/src/schemas.ts
import { z } from "zod";

export const DimensionScores = z.object({
  skills:       z.number().min(0).max(100),
  experience:   z.number().min(0).max(100),
  education:    z.number().min(0).max(100),
  cultural_fit: z.number().min(0).max(100),
});

export const CandidateEval = z.object({
  candidate_index: z.number().int().min(0),
  dimension_scores: DimensionScores,
  composite_score: z.number().min(0).max(100),
  strengths: z.array(z.string()).min(1).max(5),
  gaps:      z.array(z.string()).max(5),
  recommendation: z.enum(["Strong hire", "Consider", "Reject"]),
});

export const BatchEvalOutput = z.object({
  evaluations: z.array(CandidateEval),
});

const SkillSchema = z.object({
  name: z.string(),
  level: z.enum(["Beginner", "Intermediate", "Advanced", "Expert"]),
  yearsOfExperience: z.number().int().min(0),
});

const LanguageSchema = z.object({
  name: z.string(),
  proficiency: z.enum(["Basic", "Conversational", "Fluent", "Native"]),
});

const WorkExperienceSchema = z.object({
  company: z.string(),
  role: z.string(),
  startDate: z.string(),
  endDate: z.string(),
  description: z.string(),
  technologies: z.array(z.string()).default([]),
  isCurrent: z.boolean(),
});

const EducationEntrySchema = z.object({
  institution: z.string(),
  degree: z.string(),
  fieldOfStudy: z.string(),
  startYear: z.number().int(),
  endYear: z.number().int(),
});

const CertificationSchema = z.object({
  name: z.string(),
  issuer: z.string(),
  issueDate: z.string(),
});

const ProjectSchema = z.object({
  name: z.string(),
  description: z.string(),
  technologies: z.array(z.string()).default([]),
  role: z.string(),
  link: z.string().optional(),
  startDate: z.string(),
  endDate: z.string(),
});

const AvailabilitySchema = z.object({
  status: z.enum(["Available", "Open to Opportunities", "Not Available"]),
  type: z.enum(["Full-time", "Part-time", "Contract"]),
  startDate: z.string().optional(),
});

const SocialLinksSchema = z
  .object({
    linkedin: z.string().optional(),
    github: z.string().optional(),
    portfolio: z.string().optional(),
  })
  .catchall(z.string().optional());

export const ParsedProfileSchema = z.object({
  firstName: z.string(),
  lastName: z.string(),
  email: z.string(),
  headline: z.string(),
  bio: z.string().optional(),
  location: z.string(),
  skills: z.array(SkillSchema),
  languages: z.array(LanguageSchema).default([]),
  experience: z.array(WorkExperienceSchema),
  education: z.array(EducationEntrySchema),
  certifications: z.array(CertificationSchema).default([]),
  projects: z.array(ProjectSchema),
  availability: AvailabilitySchema,
  socialLinks: SocialLinksSchema.optional(),
});

export type CandidateEval    = z.infer<typeof CandidateEval>;
export type BatchEvalOutput  = z.infer<typeof BatchEvalOutput>;
export type ParsedProfile    = z.infer<typeof ParsedProfileSchema>;

File: backend/apps/ai/schemas.py
from pydantic import BaseModel, Field, field_validator
from typing import Literal, Optional

Recommendation = Literal["Strong hire", "Consider", "Reject"]
SkillLevel = Literal["Beginner", "Intermediate", "Advanced", "Expert"]
LanguageProficiency = Literal["Basic", "Conversational", "Fluent", "Native"]
AvailabilityStatus = Literal["Available", "Open to Opportunities", "Not Available"]
EngagementType = Literal["Full-time", "Part-time", "Contract"]


class DimensionScores(BaseModel):
    skills: float = Field(ge=0, le=100)
    experience: float = Field(ge=0, le=100)
    education: float = Field(ge=0, le=100)
    cultural_fit: float = Field(ge=0, le=100)


class CandidateEval(BaseModel):
    candidate_index: int
    dimension_scores: DimensionScores
    composite_score: float = Field(ge=0, le=100)
    strengths: list[str] = Field(min_length=1, max_length=5)
    gaps: list[str] = Field(max_length=5)
    recommendation: Recommendation


class BatchEvalOutput(BaseModel):
    evaluations: list[CandidateEval]


class Skill(BaseModel):
    name: str
    level: SkillLevel
    yearsOfExperience: int = Field(ge=0)


class Language(BaseModel):
    name: str
    proficiency: LanguageProficiency


class WorkExperience(BaseModel):
    company: str
    role: str
    startDate: str
    endDate: str
    description: str
    technologies: list[str] = []
    isCurrent: bool


class EducationEntry(BaseModel):
    institution: str
    degree: str
    fieldOfStudy: str
    startYear: int
    endYear: int


class Certification(BaseModel):
    name: str
    issuer: str
    issueDate: str


class Project(BaseModel):
    name: str
    description: str
    technologies: list[str] = []
    role: str
    link: Optional[str] = None
    startDate: str
    endDate: str


class Availability(BaseModel):
    status: AvailabilityStatus
    type: EngagementType
    startDate: Optional[str] = None


class SocialLinks(BaseModel):
    linkedin: Optional[str] = None
    github: Optional[str] = None
    portfolio: Optional[str] = None

    model_config = {"extra": "allow"}


class ParsedProfile(BaseModel):
    firstName: str
    lastName: str
    email: str
    headline: str
    bio: Optional[str] = None
    location: str
    skills: list[Skill]
    languages: list[Language] = []
    experience: list[WorkExperience]
    education: list[EducationEntry]
    certifications: list[Certification] = []
    projects: list[Project]
    availability: Availability
    socialLinks: Optional[SocialLinks] = None


class ScoringWeights(BaseModel):
    skills: float = Field(ge=0, le=1)
    experience: float = Field(ge=0, le=1)
    education: float = Field(ge=0, le=1)
    cultural_fit: float = Field(ge=0, le=1)

    @field_validator("cultural_fit")
    @classmethod
    def weights_sum_to_one(cls, v: float, info) -> float:
        data = info.data
        s = (
            data.get("skills", 0)
            + data.get("experience", 0)
            + data.get("education", 0)
            + v
        )
        if abs(s - 1.0) > 0.01:
            raise ValueError(f"weights must sum to 1.0, got {s}")
        return v


class JobRequirements(BaseModel):
    skills: list[str]
    experience_years: int
    education_level: str
    nice_to_have: list[str] = []


class Job(BaseModel):
    id: str = Field(alias="_id")
    title: str
    description: str
    requirements: JobRequirements
    scoring_weights: ScoringWeights

    model_config = {"populate_by_name": True}


class ApplicantIn(BaseModel):
    id: str = Field(alias="_id")
    parsed_profile: ParsedProfile

    model_config = {"populate_by_name": True}


# API request/response envelopes
class ScreeningRequest(BaseModel):
    run_id: str
    job: Job
    applicants: list[ApplicantIn]


class RankedResult(BaseModel):
    applicant_id: str
    rank: int
    composite_score: int
    dimension_scores: DimensionScores
    strengths: list[str]
    gaps: list[str]
    recommendation: Recommendation


class ScreeningResponse(BaseModel):
    run_id: str
    results: list[RankedResult]
